lms/services/digest.py

```diff
@@ -0,0 +1,264 @@
+from dataclasses import dataclass, field
+from typing import List, Optional
+
+from sqlalchemy import func, select
+
+from lms.models import (
+    Assignment,
+    AssignmentGrouping,
+    AssignmentMembership,
+    Course,
+    Grouping,
+    LTIRole,
+    User,
+)
+from lms.services.h_api import HAPI
+from lms.services.mailchimp import EmailRecipient, EmailSender, MailchimpService
+
+
+class DigestService:
+    def __init__(
+        self, db, helper, h_api, mailchimp_service, sender
+    ):  # pylint:disable=too-many-arguments
+        self.db = db
+        self.helper = helper
+        self.h_api = h_api
+        self.mailchimp_service = mailchimp_service
+        self.sender = sender
+
+    def send_instructor_email_digests(
+        self, h_userids, since, until, override_to_email=None
+    ):
+        annotations = self.h_api.get_annotations(h_userids, since, until)
+        unified_courses = self.helper.unified_courses(annotations)
+
+        for h_userid in h_userids:
+            unified_user = UnifiedUser.make(self.db, h_userid)
+            digest = self.helper.instructor_digest(unified_user, unified_courses)
+
+            if not digest["num_annotations"]:
+                continue
+
+            if override_to_email is None:
+                to_email = unified_user.email
+            else:
+                to_email = override_to_email
+
+            self.mailchimp_service.send_template(
+                "instructor-email-digest",
+                self.sender,
+                recipient=EmailRecipient(to_email, unified_user.display_name),
+                template_vars=digest,
+            )
+
+
+@dataclass
+class UnifiedUser:
+    """All User's for a given h_userid, unified across all ApplicationInstance's."""
+
+    users: List[User] = field(default_factory=list)
+
+    def __post_init__(self):
+        for user in self.users:
+            assert user.h_userid == self.h_userid
```

Comment:
`make` queries based on this so you wound't find this case ever.

I guess we just want to make sure that's always the case, even when not using `make`.

Reply:
Yeah, and it adds documentation of the intent. On the other hand it's already stated in the docstring so I think I'm just gonna delete this, reduce the line count